Remove DUPLICATE_PUSH and allow duplicate PUSH_PROMISE (#3309)
quicwg/base-drafts: draft-ietf-quic-http.md

3. optionally, trailing headers, if present (see {{!RFC7230}}, Section 4.1.2),
   sent as a single HEADERS frame.

> A server MAY send one or more PUSH_PROMISE (see {{frame-push-promise}}) frames

Reviewer: ```suggestion
A server MAY send one or more PUSH_PROMISE frames (see {{frame-push-promise}})
```

potential latency gain.  HTTP/3 server push is similar to what is described in
HTTP/2 {{!HTTP2}}, but uses different mechanisms.

Each server push is identified by a unique Push ID. This Push ID is used in all
PUSH_PROMISE frames (see {{frame-push-promise}}) which carries the request
headers, then included with the push stream which ultimately fulfills those

> promises.

Reviewer 2: It might be better to more explicit to say "one or more" than "all", like following, clarifying that the values MUST be the same when sent more than once.

 _This Push ID is used in one or more PUSH_PROMISE frames (see {{frame-push-promise}}) that carry the request headers, then included with the push stream which ultimately fulfills those promises. When the same Push ID is promised on multiple request streams, the decompressed request header fields MUST be byte-for-byte identical._

Reviewer 3: Good call. Done.

potential latency gain.  HTTP/3 server push is similar to what is described in
HTTP/2 {{!HTTP2}}, but uses different mechanisms.

Each server push is identified by a unique Push ID. This Push ID is used in one
or more PUSH_PROMISE frames (see {{frame-push-promise}}) that carry the request
headers, then included with the push stream which ultimately fulfills those
promises. When the same Push ID is promised on multiple request streams, the

> decompressed request header fields MUST be byte-for-byte identical.

Reviewer 3: Done.

potential latency gain.  HTTP/3 server push is similar to what is described in
HTTP/2 {{!HTTP2}}, but uses different mechanisms.

Each server push is identified by a unique Push ID. This Push ID is used in one
or more PUSH_PROMISE frames (see {{frame-push-promise}}) that carry the request
headers, then included with the push stream which ultimately fulfills those
promises. When the same Push ID is promised on multiple request streams, the

> decompressed request header set MUST contain the same set of fields in the

Reviewer: ```suggestion
decompressed request header sets MUST contain the same fields in the
```

> decompressed request header set MUST contains the same set of fields in the

Reviewer 4: ```suggestion
decompressed request header set MUST contain the same set of fields in the
```

> same order, and in each field both the name and value MUST be byte-for-byte

Reviewer: ```suggestion
same order, and both the name and the value in each field MUST be exact
```

> same order, and in each field both the name and and value MUST be byte-for-byte

Reviewer 4: ```suggestion
same order, and in each field both the name and value MUST be byte-for-byte
```

PUSH_PROMISE frame with the same push id on multiple request stream.

Ordering of a PUSH_PROMISE in relation to certain parts of the response is
important. The server SHOULD send PUSH_PROMISE frames prior to sending HEADERS
or DATA frames that reference the promised responses.  This reduces the chance
that a client requests a resource that will be pushed by the server.

When a server later fulfills a promise, the server push response is conveyed on
a push stream (see {{push-streams}}). The push stream identifies the Push ID of
the promise that it fulfills, then contains a response to the promised request
using the same format described for responses in {{request-response}}.

Due to reordering, push stream data can arrive before the corresponding
PUSH_PROMISE frame.  When a client receives a new push stream with an
as-yet-unknown Push ID, both the associated client request and the pushed
request headers are unknown.  The client can buffer the stream data in
expectation of the matching PUSH_PROMISE. The client can use stream flow control
(see section 4.1 of {{QUIC-TRANSPORT}}) to limit the amount of data a server may
commit to the pushed stream.

> If the same Push ID is promised on multiple request streams, the decompressed

Reviewer 5: I might be inclined to put this at the end of the paragraph on lines 623/624?

Reviewer 2: I think we can remove this sentence now that it is incorporated into the paragraph starting at line 605.

Sorry for me not being explicit about the move when suggesting changes to that paragraph.

Reviewer 4: agree with Kazuho but also that the the sentence should use Mike's suggestion of comparison rather than "byte-for-byte"

Reviewer 3: Done.

requests MUST conform to the requirements in Section 8.2 of {{!HTTP2}}.

The same server push can be associated with additional client requests using a

> PUSH_PROMISE frame with the same Push ID on multiple request stream.

Reviewer 6: ```suggestion
PUSH_PROMISE frame with the same Push ID on multiple request streams.
```

Ordering of a PUSH_PROMISE in relation to certain parts of the response is
important. The server SHOULD send PUSH_PROMISE frames prior to sending HEADERS
or DATA frames that reference the promised responses.  This reduces the chance
that a client requests a resource that will be pushed by the server.

When a server later fulfills a promise, the server push response is conveyed on
a push stream (see {{push-streams}}). The push stream identifies the Push ID of
the promise that it fulfills, then contains a response to the promised request
using the same format described for responses in {{request-response}}.

Due to reordering, push stream data can arrive before the corresponding
PUSH_PROMISE frame.  When a client receives a new push stream with an
as-yet-unknown Push ID, both the associated client request and the pushed
request headers are unknown.  The client can buffer the stream data in
expectation of the matching PUSH_PROMISE. The client can use stream flow control
(see section 4.1 of {{QUIC-TRANSPORT}}) to limit the amount of data a server may

> commit to the pushed stream.

Reviewer 4: Improvement suggestion: we actually have better guidance on the specific way to do flow control for the push streams in https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-6.2, so a link to that would be more useful that to QUIC Transport IMO.

PUSH_PROMISE frame that contains a larger Push ID than the client has advertised
as a connection error of H3_ID_ERROR.

A server MAY use the same Push ID in multiple PUSH_PROMISE frames. If so, the

> decompressed request header set MUST contain the same set of fields in the same

Reviewer: ```suggestion
decompressed request header sets MUST contain the same fields in the same
```

> order, and in each field both the name and and value MUST be byte-for-byte

Reviewer: ```suggestion
order, and both the name and and value in each field MUST be exact
```

matches. When a client receives a Push ID which has alread been promised, if the

> client detects a mismatch, it MUST response with a connection error of type

Reviewer: ```suggestion
and detects a mismatch, it MUST respond with a connection error of type
```

> H3_GENERAL_PROTOCOL_ERROR. Otherwise it MUST ignore the PUSH_PROMISE frame if no

Reviewer: ```suggestion
H3_GENERAL_PROTOCOL_ERROR. If the decompressed header sets match exactly, the
client MUST ignore the duplicate PUSH_PROMISE frame.
```

> mismatch is detected.

Reviewer: ```suggestion
```

PUSH_PROMISE frame that contains a larger Push ID than the client has advertised
as a connection error of H3_ID_ERROR.

A server MAY use the same Push ID in multiple PUSH_PROMISE frames. If so, the
decompressed request header sets MUST contain the same fields in the same
order, and both the name and and value in each field MUST be exact

> matches. If a client receives a Push ID that has already been promised

Reviewer 6: ```suggestion
matches. Clients SHOULD compare the request header sets for resources promised
multiple times. If a client receives a Push ID that has already been promised
```
I don't know that we need to be explicit about the discard, but I think we should encourage comparison.

and detects a mismatch, it MUST respond with a connection error of type
H3_GENERAL_PROTOCOL_ERROR. If the decompressed header sets match exactly, the

> client MUST ignore the duplicate PUSH_PROMISE frame.

Reviewer 6: ```suggestion
client SHOULD associate the pushed content with each stream on which a PUSH_PROMISE was received.
```
I don't think "MUST ignore" is what we want.  It's not an error case, it's a client optimization.

Reviewer 3: Hm. I agree that it's an error, but I'm not sure what it means to associate pushed content with the request stream that received the PUSH_PROMISE. There is only going to be a single push stream with the specified push id, of course. And as such, only a single request can be directly associated with that push stream. Of course, because of caching, other requests can be satisfied by the cached response. But that's at a layer above the streams, right?

(in chrome, I don't think we have any mapping between pushed streams and the request stream that promised the pushes so maybe we're "doing it wrong"?)

Reviewer 6: HTTP/2 associates each push with the (single) request where the server promised it.  HTTP/3 allows a push to be associated with multiple requests.

> There is only going to be a single push stream with the specified push id, of course. **And as such, only a single request can be directly associated with that push stream.**

This is the difference -- one need not follow from the other.  HTTP/3 allows a response to be promised on multiple streams, implying that you'll potentially want it as a dependency for each of those things.  That means it's intended to be possible for that push to satisfy several client requests (from different tabs, or threads, or iframes, or whatever).

Now, if the response is cacheable, you're right -- the HTTP cache will probably satisfy those other requests and it doesn't have to reach the push cache.  But the pushed response need not be cacheable; it can nonetheless be used multiple times if promised multiple times.

In WinInet, we had a concept of a navigation that was passed down from the browser, and the push cache was at the navigation level rather than the connection level.  It was part of whichever navigation included the stream that received the promise.  Hypothetically, if something were promised on several streams from different navigations, HTTP/3 would enable putting that response in both navigations' push caches.

Since Chrome can split navigations across connections, doesn't having the push cache at the connection level means you might miss that something was already pushed because it was promised on a different connection than Chrome was going to use for the request?

Reviewer 3: Hm. I think there is a difference in philosophy here. 

> HTTP/2 associates each push with the (single) request where the server promised it. HTTP/3 allows a push to be associated with multiple requests.

Totally agree with this! If a client makes requests A and B, the server can promise to push resource C on both A and B. This is clearly true.

But at the HTTP/3 layer (not at the web browsing layer) there is no notion of navigation or of subsequent requests being "associated" with earlier HTTP/3 streams (on which a resource may have been promised). Or at least I don't think there is.

My understanding of the motivation for allowing multiple pushes in HTTP/3 is not because we want to allow multiple subsequent requests for the same resources to be associated with the same pushed stream. Instead it was because HTTP/2 is in-order and HTTP/3 is not. In HTTP/2 if a client requested A and B and the server knew that both referenced resource C, it could push C a single time and be confident that requests initiated by the reference to C in the bodies of A and B would both come *after* the promise of C. However in HTTP/3 this is not the case. Since the promise for C is delivered on the request stream there is no ordering guarantee that the push (on A) will arrive at the client before the body of B (which contains the reference to C) arrives. As such, if the server only wants to push C once, it needs to make sure both promises are know.

But this, to me, is orthogonal to the issue of associating multiple subsequent requests with the same push data down at the HTTP/3 layer because each of these requests are somehow associated with the original requests A and B.

Am I thinking about this the wrong way?

> Since Chrome can split navigations across connections, doesn't having the push cache at the connection level means you might miss that something was already pushed because it was promised on a different connection than Chrome was going to use for the request?

While Chrome does store pushes in a per-connection cache (which is really just the pushed streams themselves) it also has a "global" index of "promises by URL" which allows a request which would, by default, go on one connection to be satisfied by a promise from another connection (which might otherwise be going away, for example).

Reviewer 6: Yes, I think we've found a conceptual difference.  You're looking at the promise purely as a way to forestall requests for that resource, and the fact that it happens to be on a given stream is only because the server thinks there might be something on that stream where you need to know about the push first.  But handling of push in Chrome is global -- if you request A and B, I promise X on A, and B references X, you'll use the pushed object if you know about it.  I agree, in this implementation style, the first PUSH_PROMISE frame to arrive carries all the information you need, and any subsequent ones are basically irrelevant.

But RFC7540 says that "Pushed responses are always associated with an explicit request from the client."  How (or whether, obviously!) that association is used is implementation-dependent, but I think for consistency's sake we need to provide that association in the document.  (As far as I know, the only thing H2 explicitly used this association for internally was for prioritization -- pushed streams depend on the associated stream.)

Do we need a separate editorial issue to expand on this a bit in the document, or is it there and we're just reading through different filters?

Reviewer 3: > Yes, I think we've found a conceptual difference. You're looking at the promise purely as a way to forestall requests for that resource, and the fact that it happens to be on a given stream is only because the server thinks there might be something on that stream where you need to know about the push first. But handling of push in Chrome is global -- if you request A and B, I promise X on A, and B references X, you'll use the pushed object if you know about it. I agree, in this implementation style, the first PUSH_PROMISE frame to arrive carries all the information you need, and any subsequent ones are basically irrelevant.
> 
> But RFC7540 says that "Pushed responses are always associated with an explicit request from the client." How (or whether, obviously!) that association is used is implementation-dependent, but I think for consistency's sake we need to provide that association in the document. (As far as I know, the only thing H2 explicitly used this association for internally was for prioritization -- pushed streams depend on the associated stream.)

Interesting! I read the language in RFC7540 differently, I think. It says "Pushed responses are always associated with an explicit request from the client.  The PUSH_PROMISE frames sent by the server are sent on that explicit request's stream. " To me, this language exists to prevent unilateral pushes by the server. In other words, if there are no client-initiated requests outstanding, the server may not push any resources. 

But if I understand you correctly, you read it as enabling implementations to make use of this association for more than simply preventing unilateral pushes. (Presumably this association could be used by an implementation to "hide" the pushed resource from a subsequent request not associated with the original request, but to "show" it to a subsequent request that *was* associated with the original request? I would not have thought this was "a thing".) Fascinating.

> Do we need a separate editorial issue to expand on this a bit in the document, or is it there and we're just reading through different filters?

This is a good question :) I think some editorial exposition on this front would be welcome! I would be very curious to hear what other folks think of this issue.

In any case, backing up to the change you proposed:

> client SHOULD associate the pushed content with each stream on which a PUSH_PROMISE was received.

I think I'm comfortable with this text even though I'm slightly uneasy about what "associate" means here. But that's pretty orthogonal to this PR since it existed in the DUPLICATE_PUSH language as well, so I'll go ahead and apply your suggestion.

Thanks for the great discussion!

Allowing duplicate references to the same Push ID is primarily to reduce
duplication caused by concurrent requests.  A server SHOULD avoid reusing a Push
ID over a long period.  Clients are likely to consume server push responses and
not retain them for reuse over time.  Clients that see a PUSH_PROMISE that uses

> a Push ID that they have since consumed and discarded are forced to ignore the

Reviewer 6: ```suggestion
a Push ID that they have already consumed and discarded are forced to ignore the
```

PUSH_PROMISE frame that contains a larger Push ID than the client has advertised
as a connection error of H3_ID_ERROR.

A server MAY use the same Push ID in multiple PUSH_PROMISE frames. If so, the

> decompressed request headers MUST be idential.  A client MUST treat receipt of a

Reviewer 4: ```suggestion
decompressed request headers MUST be identical.  A client MUST treat receipt of a
```

Reviewer 4: maybe a x-ref to the definition of identical would help too

Reviewer 3: Modified to include the same language as on line 605. Hope this helps!

Push ID which has already been promised with different headers than the previous

> promise as a connection error of type H3_ID_ERROR.

Reviewer 6: MUST-error requires that the client keep the request headers for the lifetime of the connection in case it ever sees that Push ID again.  Instead, I think we want to say that it MUST error if it detects a mismatch, and MUST ignore the new PUSH_PROMISE if it has not retained the previous request header set.

Reviewer 2: > MUST-error requires that the client keep the request headers for the lifetime of the connection in case it ever sees that Push ID again.

As stated in https://github.com/quicwg/base-drafts/issues/3275#issuecomment-566379821, it is my understanding that an endpoint can determine if the PUSH_PROMISE refers to a new Push ID or a Push ID it already has received, and therefore my preference goes to stating "MUST error if there is a mismatch, or ignore the PUSH_PROMISE frame."

Reviewer 4: No comment about the MUSTiness but H3_ID_ERROR is not appropriate for this error condition IMO. Nor is it a frame error because all the syntax was correct. The closest comparison is malformed request, which elicits a H3_GENERAL_PROTOCOL_ERROR. Alternatively connection-level H3_REQUEST_REJECTED could fit but might need some other changes in the wider document to make sure this case is accommodated for.

Reviewer 3: I've updated the text to use the "MUST error if mismatch detected" language.
